File: oldfoundation/trie.cpp
#include<iostream>
#include<unordered_map>

using namespace std;
class node
{
    public:
    char data;
    bool end;
    unordered_map<char,node*>children;

};
class trie
{
    private:
    node* root;
    int words;
    int nodes;

    bool search(node* temp,string word,int idx){
        if(idx==word.length()){
            return temp->end;
        }
        char ch=word[idx];
        if(temp->children.count(ch)==0){
            return false;
        }else{
          node* child=temp->children[ch];
          return search(temp->children[ch],word,idx+1);

        }
    }
    void add(node* temp,string word,int idx){
        if(idx==word.length()){
            words++;
            temp->end=true;
            return;
        }

        char ch=word[idx];
        if(temp->children.count(ch)==0){
            nodes++;

            node* child=new node();
            child->data=ch;
            child->end=false;
            temp->children[ch]=child;
            add(child,word,idx+1);
        }else{
            node* child=temp->children[ch];
            add(child,word,idx+1);
        }
    }
    void displayWords(node* temp,string wsf){
        if(temp->end){
            cout<<wsf<<endl;
        }
        for(auto itr=temp->children.begin();itr!=temp->children.end();itr++){
          displayWords(itr->second,wsf+itr->first);
        }
    }
    
    void displayTrie(node* temp){
        cout<<temp->data<<"->";
        for(auto itr=temp->children.begin();itr!=temp->children.end();itr++){
            cout<<itr->first<<",";//itr->second->end?"*":",";
        }
        cout<<"."<<endl;
         for(auto itr=temp->children.begin();itr!=temp->children.end();itr++){
            displayTrie(itr->second);
        }

    }
    void remove(node* temp,string word,int idx){
        if(idx==word.length()){
         words--;
         temp->end=false;
         return;
        }
        char ch=word[idx];
        node* child=temp->children[ch];
        remove(temp->children[ch],word,idx+1);
        if(child->end==false && child->children.size()==0){
            temp->children.erase(ch);
            nodes--;
            delete child;
        }
    }
    public :
    trie(){
        root=new node();
        root->data='$';
        root->end=false;
        words=0;
        nodes=1;
    }
    bool search(string words){
         return search(root,words,0);
           
    }
    void add(string word){
        if(search(word)==false){
            add(root,word,0);
        }
    }
    void displayWords(){
         displayWords(root," ");
    }
    
    void remove(string word){
        remove(root,word,0);//dd
    }
    void displayTrie(){
        displayTrie(root);
    }
};
int main(int argc,char** argv){
    trie t;
    t.add("are");
    t.add("art");
    t.add("an");
    t.add("and");
    t.add("as");
    t.add("ask");
    t.add("see");
    t.add("sea");
    t.add("seen");
      t.displayWords();

    cout<<t.search("artist")<<t.search("se")<<t.search("seen")<<endl;
   t.remove("as");
   t.remove("ask");
   t.displayTrie();
}

File: Trie-textprocessing/Trie.java
import java.util.*;

public class Trie{
     static class Node{      //Node have three property
        char data;
        HashMap<Character,Node> children;  //Right bcz parent is one 
        boolean eow;  //end of word
        Node(char data){
            this.data=data;
            this.children=new HashMap<>();
            this.eow=false;
        }
    }
    public static void add(Node node,String word){
        if(word.length()==0){
            node.eow=true;
            return ;
        }
        char ch=word.charAt(0);
        String row=word.substring(1);
        Node child=node.children.get(ch);  //ch-key & child-value(address of next character)
        if(child==null){        //if child not present then add child
            child=new Node(ch);
            node.children.put(ch,child);
            add(child, row);
        }else{
            add(child, row);
        }
        
    }
    public static boolean search(Node node,String word){
       if(word.length()==0){
           return node.eow;
       }
       
       char ch=word.charAt(0);
       String row=word.substring(1); //rest of word
       Node child=node.children.get(ch);
       if(child==null){
         return false;
       }else{
         return search(child,row);
       }
    }
    //V.Important
    public static void remove(Node node,String word){
        if(word.length()==0){
            node.eow=false;
            return;
        }
        char ch=word.charAt(0);
        String row=word.substring(1);
        Node child=node.children.get(ch);

      if(child==null){
          return;
      }else{
          remove(child, row);               //remove in post order
          if(child.eow==false && child.children.size()==0){       //Removing character is not end part character 
                                        //of other word && also removing character is not a mid part of other word
              node.children.remove(ch);  //remove key from above node of below Node(child).
          }
      }
    }
    public static void display(Node node){
        String str=node.data+"->";
        ArrayList<Character> keys=new ArrayList<>(node.children.keySet()); // use to run loop
       for(Character key:keys){
           Node child=node.children.get(key);
           str+=child.data+(child.eow==true?"*":"")+",";
       }
       str+=".";
       System.out.println(str);
       for(Character key:keys){
        Node child=node.children.get(key);
        display(child);
       }
    }
    public static void displayAllwords(Node node,String wsf){
        if(node.eow==true){
        System.out.println(wsf);      //No need to write return;
        }
       ArrayList<Character> keys=new ArrayList<>(node.children.keySet());
       for(Character key:keys){
           Node child=node.children.get(key);
           displayAllwords(child,wsf+key);
       }
    }
    static int[] getlps(String pat){ //Pi table
        int i=1;
        int j=0;
        
        int[]lps=new int[pat.length()];
        lps[0]=0;
        while(i<pat.length()){
            if(pat.charAt(i)==pat.charAt(j)){
                lps[i]=j+1;
                i++;
                j++;
            }else if(j==0){      //eg-aaaay
                    lps[i]=0;
                    i++;
            }else{
                j=lps[j-1];
                }
        }
        return lps;
    }

    
    static void KMP(String src,String pat){
        int[] lps=getlps(pat);         //longest proper suffix
        int i=0;       //traverse on source
        int j=0;      //traverse on  pattern
        while(i<src.length()){
            if(src.charAt(i)==pat.charAt(j)){
                i++;
                j++;
                if(j==pat.length()){
                    System.out.println("Found at "+(i-pat.length()));
                    j=lps[j-1];     //not j=0;(test case)(String src="abhnramramramnkcjjcnramramramjcrahulnjenrahulwramramramccccrahec";
                }                               //String pat="ramram";)
            }else if(j>0){
                 j=lps[j-1];
            }else{               //if(j==0)
                 i++;
               }
            
        }
    }

  
    public static void main(String[] args){
        Node root=new Node('$');
        add(root, "are");
        add(root, "art");
        add(root, "an");
        add(root, "and");
        add(root, "ant");
        add(root, "as");
        add(root, "ask");
        add(root, "see");
        add(root, "sea");
        add(root, "seen");
        System.out.println(search(root,"art"));
        // remove(root, "seen");
        display(root);
        System.out.println("------------");
        // displayAllwords(root, "");
        // remove(root, "art");
        // System.out.println("------------");
        // display(root);
        // System.out.println("------------");
        // displayAllwords(root, "");
        // String src="abhnramramramnkcjjcnramramramjcrahulnjenrahulwramramramccccrahec";
        // String pat="ramram";         
        // KMP2(src, pat);

    }
}